### PR TITLE
feat(operator): add migration orchestration controller

### PR DESCRIPTION
**Pull request by Siddhant-K-code**
## Summary

Adds the Go controller that coordinates OpenFGA datastore migrations before application rollout.

- Derives the desired OpenFGA version and migration configuration from the target Deployment.
- Creates and observes migration Jobs, then records completed versions in a status ConfigMap.
- Handles retries, stale Jobs, failure conditions, and upgrade sequencing.
- Rejects unmanaged Job and ConfigMap name collisions instead of overwriting unrelated resources.
- Uses ownership checks plus UID and resource-version preconditions for safe deletion.
- Includes focused controller coverage for success, failure, collision, and race scenarios.

## Stack

Native stack 351. [Open #345 to view the GitHub stack map](https://github.com/openfga/helm-charts/pull/345). Review bottom to top:

1. #345: External migration hooks, base `main`
2. **#346: Migration controller (this PR), base #345**
3. #347: Operator Helm chart, base #346
4. #348: OpenFGA chart integration, base #347
5. #349: Operator CI and upgrade E2E, base #348
6. #350: Migration architecture ADRs, base #349

Replaces closed #337 and carries the controller portion of source #331. Helm packaging and chart integration remain in later layers.

## Attribution

The 24 original controller commits are replayed with [Ed Milic (@emilic)](https://github.com/emilic) as their actual Git author, preserving their original author dates, messages, and commit boundaries. GitHub links the replayed commits to `@emilic`.

## Validation

- `make vet test` in `operator`
- Exact ten-path controller scope
- Reviewed #337 tree parity
- `git diff --check`

### CODE REVIEW

File: operator/internal/controller/migration_controller.go [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+
+	if !isOperatorManaged(job) && !isOwnedByDeployment(job, deployment) {
+		return ctrl.Result{}, fmt.Errorf(
+			"migration job %s/%s exists but is neither managed by %s nor owned by Deployment %s",
+			job.Namespace,
+			job.Name,
+			LabelManagedByValue,
+			deployment.Name,
+		)
+	}
```
[the remark this answers is not in the document]
Comment (Siddhant-K-code) [reply operator/internal/controller/migration_controller.go]: Updated the existing Job guard so a Job is trusted or deleted only when it has the `openfga-operator` managed-by label and is owned by the exact target Deployment. Added focused label-only, owner-only, both, and neither cases; only the both case proceeds to delete the trusted stale Job.

File: operator/internal/controller/migration_controller.go [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+	cmName := migrationConfigMapName(req.Name)
+	err = r.Get(ctx, types.NamespacedName{Name: cmName, Namespace: req.Namespace}, configMap)
+
+	currentVersion := ""
+	if err == nil {
+		if isOperatorManaged(configMap) {
+			currentVersion = configMap.Data["version"]
+		} else {
+			logger.Info("ignoring unmanaged migration status ConfigMap", "configMap", cmName)
+		}
```
[the remark this answers is not in the document]
Comment (Siddhant-K-code) [reply operator/internal/controller/migration_controller.go]: Changed reconciliation to return an explicit collision error immediately when the same-named migration-status ConfigMap is not operator-managed. The focused test now verifies that no migration Job is created and the colliding ConfigMap remains unchanged.

**Review by SoulPancake — COMMENTED** [review]
- `ensureDeploymentScaled` enforces `desired-replicas` on every reconcile, not just after migration. Any `kubectl scale` is instantly reverted (the scale event itself triggers the reconcile), and with autoscaling blocked in operator mode that's the only scaling left. Also snaps back the live replica count the chart's `lookup` deliberately preserves, right after the Job succeeds. Make it a one-shot gate — only scale when replicas is 0 — and add a test that a non-zero live count is left alone.
- Stale-version/legacy-hook Jobs are deleted with background propagation while possibly still running, and the replacement lands ~5s later → two concurrent `openfga migrate` runs while the old pod terminates. Foreground propagation fixes it: the Job lingers until pods are gone, so recreation waits naturally.

Nits: `cmd/main.go` fails `gofmt -l`; module path should be `github.com/openfga/helm-charts/operator` (operator lives here per ADR-003) unless extraction is planned; `updateMigrationStatus` overwrites `existing.Labels`

File: operator/internal/controller/helpers.go [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+	existing.Labels = cm.Labels
+	existing.OwnerReferences = cm.OwnerReferences
+	if updateErr := c.Update(ctx, existing); updateErr != nil {
+		return fmt.Errorf("updating migration status ConfigMap: %w", updateErr)
+	}
+	return nil
+}
+
+func deleteMigrationJob(ctx context.Context, c client.Client, job *batchv1.Job) error {
+	propagation := metav1.DeletePropagationBackground
```
[the remark this answers is not in the document]
Comment (Siddhant-K-code) [reply operator/internal/controller/helpers.go]: Fixed in `d0e22e5d`. Job cleanup now uses foreground propagation with UID/resource-version preconditions, and reconciliation waits while deletion is in progress so a replacement cannot overlap a terminating migration Pod.

File: operator/README.md [hunk 4]
```diff
@@ -0,0 +1,134 @@
+# OpenFGA Operator
+
+A Kubernetes operator that manages database migrations for OpenFGA deployments. Instead of relying on Helm hooks and init containers, the operator watches OpenFGA Deployments, detects version changes, and orchestrates migrations as regular Jobs.
+
+This is **Stage 1** of the operator — focused solely on migration orchestration. See [ADR-001](../docs/adr/001-adopt-openfga-operator.md) for the full roadmap.
```
[the remark this answers is not in the document]
Comment (Siddhant-K-code) [reply operator/README.md]: Fixed in `d0e22e5d`. The link was removed from this controller-only layer because ADR-001 is introduced later in the stack; the README now contains only targets available in this tree.

File: operator/README.md [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+
+### Docker Image
+
+```bash
+docker build -t openfga/openfga-operator:dev .
+```
+
+## Local Testing
+
+Integration test values and instructions are in [`tests/`](tests/). Three scenarios are provided:
```
[the remark this answers is not in the document]
Comment (Siddhant-K-code) [reply operator/README.md]: Fixed in `d0e22e5d`. The integration-only quick start and absent test/value references were removed from this controller layer; the usable deployment guidance belongs to the later chart-integration layer where its dependencies exist.

File: operator/internal/controller/migration_controller.go [hunk 6]
```diff
@@ -0,0 +1,373 @@
...
+func clearMigrationFailedCondition(deployment *appsv1.Deployment) {
+	for i, c := range deployment.Status.Conditions {
+		if c.Type == "MigrationFailed" {
+			deployment.Status.Conditions[i].Status = corev1.ConditionFalse
+			deployment.Status.Conditions[i].LastTransitionTime = metav1.Now()
+			deployment.Status.Conditions[i].Reason = "MigrationSucceeded"
+			deployment.Status.Conditions[i].Message = "Migration completed successfully."
+			return
+		}
+	}
```
[the remark this answers is not in the document]
Comment (Siddhant-K-code) [reply operator/internal/controller/migration_controller.go:578]: Fixed in `d0e22e5d`. Clearing `MigrationFailed` is now idempotent when status, reason, and message already match, so repeated reconciles do not patch status or refresh `LastTransitionTime`.

File: operator/internal/controller/helpers.go [hunk 7]
```diff
@@ -0,0 +1,357 @@
...
+	releaseInstance := deploymentLabels[LabelInstance]
+	if releaseInstance == "" || jobLabels[LabelInstance] != releaseInstance {
+		return false
+	}
+
+	for _, identityLabel := range []string{LabelName, LabelPartOf, LabelComponent, LabelHelmChart} {
+		if expected := deploymentLabels[identityLabel]; expected != "" && jobLabels[identityLabel] != expected {
+			return false
+		}
+	}
```
[the remark this answers is not in the document]
Comment (Siddhant-K-code) [reply operator/internal/controller/helpers.go]: Fixed in `d0e22e5d`. Legacy adoption no longer compares the versioned `helm.sh/chart` label; it retains strict checks for stable release identity and allowed hook events, with cross-version and mismatch coverage.

File: operator/cmd/main.go [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+		name  string
+		value int
+		max   int
+	}{
+		{"backoff-limit", backoffLimit, math.MaxInt32},
+		{"active-deadline-seconds", activeDeadline, math.MaxInt32},
+		{"ttl-seconds-after-finished", ttlAfterFinished, math.MaxInt32},
+	} {
+		if v.value < 0 || v.value > v.max {
+			fmt.Fprintf(os.Stderr, "invalid value for --%s: must be between 0 and %d\n", v.name, v.max)
```
[the remark this answers is not in the document]
Comment (Siddhant-K-code) [reply operator/cmd/main.go]: Fixed in `d0e22e5d`. `activeDeadlineSeconds` now requires at least 1, while zero remains valid for `backoffLimit` and `ttlSecondsAfterFinished`; boundary tests cover each flag.

File: operator/cmd/main.go [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+
+	ctrl.SetLogger(zap.New(zap.UseFlagOptions(&opts)))
+	logger := ctrl.Log.WithName("setup")
+
+	// Fall back to the pod's namespace when no explicit scope is set.
+	if watchNamespace == "" {
+		if podNS, ok := os.LookupEnv("POD_NAMESPACE"); ok && podNS != "" {
+			watchNamespace = podNS
+			logger.Info("defaulting watch scope to pod namespace", "namespace", podNS)
+		}
```
[the remark this answers is not in the document]
Comment (Siddhant-K-code) [reply operator/cmd/main.go]: Fixed in `d0e22e5d`. Namespace resolution now uses the explicit flag, then `POD_NAMESPACE`, then the service-account namespace file, and fails clearly if none yields a namespace instead of widening the cache scope.

File: operator/internal/controller/migration_controller.go [hunk 10]
```diff
@@ -0,0 +1,373 @@
...
+	// to users within seconds instead of minutes.
+	if isJobConditionTrue(job, batchv1.JobFailed) || isJobConditionTrue(job, batchv1.JobFailureTarget) {
+		logger.Info("migration job failed, will delete and retry", "job", jobName, "version", desiredVersion)
+
+		// Set condition so kubectl describe shows the failure.
+		statusPatch := client.MergeFrom(deployment.DeepCopy())
+		setMigrationFailedCondition(deployment, desiredVersion)
+		if patchErr := r.Status().Patch(ctx, deployment, statusPatch); patchErr != nil {
+			logger.Error(patchErr, "failed to set MigrationFailed condition")
+		}
```
[the remark this answers is not in the document]
Comment (Siddhant-K-code) [reply operator/internal/controller/migration_controller.go]: Fixed in `d0e22e5d`. A failed Deployment status patch now returns immediately, before retry metadata or Job deletion; regression coverage verifies the failed Job remains available for the next reconciliation.

**Comment by Siddhant-K-code** [comment]
Addressed the controller review in `d0e22e5d`.

- Replica restoration is now a one-shot zero-replica gate with optimistic locking; non-zero live counts are left untouched.
- Migration Job deletion uses foreground propagation and waits for termination before replacement.
- Go formatting is clean, and the module path is now `github.com/openfga/helm-charts/operator`.
- Migration status updates preserve unrelated ConfigMap labels while enforcing required labels.

The same commit also fixes the namespace-scope fallback, active-deadline validation, status-patch durability, idempotent failure-condition clearing, and cross-chart-version legacy Job adoption. Full tests, race tests, vet, and build pass.

File: operator/internal/controller/migration_controller.go [hunk 11]
```diff
@@ -0,0 +1,390 @@
...
+		currentVersion = configMap.Data["version"]
+	} else if !apierrors.IsNotFound(err) {
+		return ctrl.Result{}, fmt.Errorf("getting migration status: %w", err)
+	}
+
+	// 5. If versions match, ensure Deployment is scaled up and return.
+	if currentVersion == desiredVersion {
+		logger.V(1).Info("migration up to date", "version", desiredVersion)
+		statusPatch := client.MergeFrom(deployment.DeepCopy())
```
[the remark this answers is not in the document]
Comment (Siddhant-K-code) [reply operator/internal/controller/migration_controller.go]: Fixed in `3fd70630`. Every Deployment status patch that can replace conditions now uses optimistic locking, while unchanged failure clearing remains an idempotent no-op; concurrency tests verify condition preservation.

File: operator/internal/controller/migration_controller.go [hunk 12]
```diff
@@ -0,0 +1,390 @@
...
+	// JobFailureTarget is set as soon as the Job controller decides the Job
+	// will fail (backoff limit reached, deadline exceeded, etc.); JobFailed
+	// only flips after pods finish terminating, which can take BackoffLimit ×
+	// ActiveDeadlineSeconds. Treating either as "failed" surfaces the failure
+	// to users within seconds instead of minutes.
+	if isJobConditionTrue(job, batchv1.JobFailed) || isJobConditionTrue(job, batchv1.JobFailureTarget) {
+		logger.Info("migration job failed, will delete and retry", "job", jobName, "version", desiredVersion)
+
+		// Set condition so kubectl describe shows the failure.
+		statusPatch := client.MergeFrom(deployment.DeepCopy())
```
[the remark this answers is not in the document]
Comment (Siddhant-K-code) [reply operator/internal/controller/migration_controller.go]: Fixed in `3fd70630`. Setting `MigrationFailed` now uses a resource-version optimistic-lock patch, so concurrent Deployment status changes conflict and retry instead of being overwritten.

File: operator/internal/controller/helpers.go [hunk 13]
```diff
@@ -0,0 +1,362 @@
...
+					Containers: []corev1.Container{
+						{
+							Name:            "migrate-database",
+							Image:           mainContainer.Image,
+							Args:            []string{"migrate"},
+							Env:             mainContainer.Env,
+							EnvFrom:         mainContainer.EnvFrom,
+							Resources:       mainContainer.Resources,
+							VolumeMounts:    mainContainer.VolumeMounts,
+							SecurityContext: mainContainer.SecurityContext,
```
[the remark this answers is not in the document]
Comment (Siddhant-K-code) [reply operator/internal/controller/helpers.go:177]: Fixed in `3fd70630`. Migration Jobs now copy the selected OpenFGA container's exact `ImagePullPolicy`; coverage includes Never, Always, IfNotPresent, defaulting, and `latest` with `Never`.

File: operator/README.md [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+
+## Annotations
+
+The operator reads these annotations from the OpenFGA Deployment:
+
+| Annotation | Description |
+|------------|-------------|
+| `openfga.dev/migration-enabled` | Must be `"true"` for the operator to manage migrations. Deployments without this annotation are ignored. |
+| `openfga.dev/desired-replicas` | The replica count to restore after migration succeeds when the Deployment is still at zero replicas. |
+| `openfga.dev/migration-service-account` | The ServiceAccount to use for migration Jobs. Defaults to the Deployment's SA. |
```
[the remark this answers is not in the document]
Comment (Siddhant-K-code) [reply operator/README.md:91]: Fixed in `3fd70630`. The annotations table now documents `openfga.dev/container-name`, including its default and how it selects a non-default OpenFGA container.

File: operator/internal/controller/helpers.go [hunk 15]
```diff
@@ -0,0 +1,362 @@
...
+					UID:                deployment.UID,
+					Controller:         ptr.To(true),
+					BlockOwnerDeletion: ptr.To(true),
+				},
+			},
+		},
+		Spec: batchv1.JobSpec{
+			BackoffLimit:            ptr.To(backoffLimit),
+			ActiveDeadlineSeconds:   ptr.To(activeDeadlineSeconds),
+			TTLSecondsAfterFinished: ptr.To(ttlSecondsAfterFinished),
```
[the remark this answers is not in the document]
Comment (Siddhant-K-code) [reply operator/internal/controller/helpers.go]: Fixed in `3fd70630`. New Jobs are no longer TTL-eligible before their terminal result is persisted; reconciliation records the result first, then arms TTL idempotently with optimistic locking, including TTL=0 and restart coverage.

File: operator/internal/controller/migration_controller.go [hunk 16]
```diff
@@ -0,0 +1,390 @@
...
+		if patchErr := r.Patch(ctx, deployment, patch); patchErr != nil {
+			return ctrl.Result{}, fmt.Errorf("persisting retry-after annotation: %w", patchErr)
+		}
+
+		// Delete the failed Job so a fresh one is created on the next reconcile.
+		if delErr := deleteMigrationJob(ctx, r.Client, job); delErr != nil && !apierrors.IsNotFound(delErr) {
+			if apierrors.IsConflict(delErr) {
+				return ctrl.Result{RequeueAfter: time.Second}, nil
+			}
+			return ctrl.Result{}, fmt.Errorf("deleting failed migration job: %w", delErr)
```
[the remark this answers is not in the document]
Comment (Siddhant-K-code) [reply operator/internal/controller/migration_controller.go]: Fixed in `3fd70630`. Failed Jobs are retained through the later of the retry cooldown and configured diagnostic window, then foreground-deleted with preconditions before any replacement is created.

File: operator/internal/controller/migration_controller.go [hunk 17]
```diff
@@ -0,0 +1,390 @@
...
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	// 2. Skip if migration is not opted-in via annotation.
+	if len(deployment.Annotations) == 0 || deployment.Annotations[AnnotationMigrationEnabled] != "true" {
+		logger.V(1).Info("migration not enabled for this deployment, skipping")
+		return ctrl.Result{}, nil
+	}
```
[the remark this answers is not in the document]
Comment (Siddhant-K-code) [reply operator/internal/controller/migration_controller.go]: Fixed in `3fd70630`. Reconcile now rechecks both required labels plus migration enablement before any mutation, covering owner-enqueued Job and ConfigMap events after a Deployment leaves scope.

File: operator/internal/controller/migration_controller.go [hunk 18]
```diff
@@ -0,0 +1,390 @@
...
+				LabelManagedByValue,
+			)
+		}
+		currentVersion = configMap.Data["version"]
+	} else if !apierrors.IsNotFound(err) {
+		return ctrl.Result{}, fmt.Errorf("getting migration status: %w", err)
+	}
+
+	// 5. If versions match, ensure Deployment is scaled up and return.
+	if currentVersion == desiredVersion {
```
[the remark this answers is not in the document]
Comment (Siddhant-K-code) [reply operator/internal/controller/migration_controller.go:115]: Fixed in `3fd70630`. A managed migration-status ConfigMap now has its controller ownership repaired to the current Deployment with optimistic locking before the current-version early return; conflicts retry and a successful repair does not rerun migration.

**Comment by Siddhant-K-code** [comment]
Closing this stack for now. We're going to focus first on updating, reviewing, and fixing #309.
